File: backend/src/utils/otpService.js
const Otp = require('../models/Otp');

const generateOtpCode = () =>
  Math.floor(100000 + Math.random() * 900000).toString();

const sendOtpSms = async (mobile, code) => {
  if (process.env.TWILIO_ACCOUNT_SID && process.env.TWILIO_AUTH_TOKEN) {
    const twilio = require('twilio')(
      process.env.TWILIO_ACCOUNT_SID,
      process.env.TWILIO_AUTH_TOKEN
    );
    await twilio.messages.create({
      body: `Your PujaSetu OTP is ${code}. Valid for ${process.env.OTP_EXPIRY_MINUTES || 10} minutes.`,
      from: process.env.TWILIO_PHONE_NUMBER,
      to: `+91${mobile}`,
    });
    return;
  }
  console.log(`[DEV OTP] ${mobile}: ${code}`);
};

const createAndSendOtp = async (mobile) => {
  const code =
    process.env.NODE_ENV === 'development' && process.env.DEV_OTP_BYPASS === 'true'
      ? process.env.DEV_OTP_CODE || '123456'
      : generateOtpCode();

  const expiresAt = new Date(
    Date.now() + (parseInt(process.env.OTP_EXPIRY_MINUTES, 10) || 10) * 60 * 1000
  );

  await Otp.deleteMany({ mobile });
  await Otp.create({ mobile, code, expiresAt });
  await sendOtpSms(mobile, code);

  return { expiresAt, devMode: process.env.DEV_OTP_BYPASS === 'true' };
};

const verifyOtp = async (mobile, code) => {
  const record = await Otp.findOne({ mobile }).sort({ createdAt: -1 });
  if (!record) return { valid: false, message: 'OTP not found. Request a new one.' };
  if (record.expiresAt < new Date()) return { valid: false, message: 'OTP expired' };
  if (record.attempts >= 5) return { valid: false, message: 'Too many attempts' };

  const devBypass =
    process.env.DEV_OTP_BYPASS === 'true' && code === (process.env.DEV_OTP_CODE || '123456');

  if (!devBypass && record.code !== code) {
    record.attempts += 1;
    await record.save();
    return { valid: false, message: 'Invalid OTP' };
  }

  await Otp.deleteMany({ mobile });
  return { valid: true };
};

module.exports = { createAndSendOtp, verifyOtp };

File: backend/src/main/java/com/pujasetu/service/OtpService.java
package com.pujasetu.service;

import com.pujasetu.config.AppProperties;
import com.pujasetu.model.OtpRecord;
import com.pujasetu.repository.OtpRepository;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Service;

import java.time.Instant;
import java.util.List;
import java.util.Random;

@Slf4j
@Service
@RequiredArgsConstructor
public class OtpService {

    private final OtpRepository otpRepository;
    private final AppProperties appProperties;
    private final Random random = new Random();

    public record OtpResult(Instant expiresAt, boolean devMode) {}

    public OtpResult createAndSendOtp(String mobile) {
        String code = appProperties.getOtp().isDevBypass()
                ? appProperties.getOtp().getDevCode()
                : String.format("%06d", 100000 + random.nextInt(900000));

        Instant expiresAt = Instant.now().plusSeconds(appProperties.getOtp().getExpiryMinutes() * 60L);

        otpRepository.deleteByMobile(mobile);
        otpRepository.save(OtpRecord.builder()
                .mobile(mobile)
                .code(code)
                .expiresAt(expiresAt)
                .build());

        sendSms(mobile, code);
        return new OtpResult(expiresAt, appProperties.getOtp().isDevBypass());
    }

    public record VerifyResult(boolean valid, String message) {}

    public VerifyResult verifyOtp(String mobile, String code) {
        List<OtpRecord> records = otpRepository.findByMobileOrderByExpiresAtDesc(mobile);
        if (records.isEmpty()) {
            return new VerifyResult(false, "OTP not found. Request a new one.");
        }

        OtpRecord record = records.getFirst();
        if (record.getExpiresAt().isBefore(Instant.now())) {
            return new VerifyResult(false, "OTP expired");
        }
        if (record.getAttempts() >= 5) {
            return new VerifyResult(false, "Too many attempts");
        }

        boolean devBypass = appProperties.getOtp().isDevBypass()
                && code.equals(appProperties.getOtp().getDevCode());

        if (!devBypass && !record.getCode().equals(code)) {
            record.setAttempts(record.getAttempts() + 1);
            otpRepository.save(record);
            return new VerifyResult(false, "Invalid OTP");
        }

        otpRepository.deleteByMobile(mobile);
        return new VerifyResult(true, "OK");
    }

    private void sendSms(String mobile, String code) {
        var twilio = appProperties.getTwilio();
        if (twilio.getAccountSid() != null && !twilio.getAccountSid().isBlank()) {
            log.info("Twilio SMS would be sent to +91{}", mobile);
            return;
        }
        log.info("[DEV OTP] {}: {}", mobile, code);
    }
}
